File: desktop/src/utils/deckFormats.js
// Spec E2 §3 — Decklisten-Formate YDK, YDKE und Textliste: lesen, schreiben, erkennen.
// ZWILLING: android/app/src/main/java/com/example/yugiohscanner/ml/DeckFormats.kt. Beide laufen gegen
// docs/fixtures/decks/formats.json. Wer eine Seite aendert, aendert beide.

export const YDKE_INVALID = 'Kein gültiger YDKE-Link';
export const NOTHING_RECOGNIZED = 'Keine Deckliste erkannt';
export const YDK_HEADER = '#created by YGO Card Manager';

const MAX_UINT32 = 4294967295;
// Standard-Base64 mit Auffuellung; leerer Block erlaubt. Beide Zwillinge pruefen mit DIESEM Muster, weil atob und
// java.util.Base64 unterschiedlich nachsichtig sind.
const BASE64 = /^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/;
const HEADING = /^(main|extra|side)(?:\s+deck)?\s*(?::|\([0-9]+\))?$/i;
// [0-9] statt \d und Unicode-\s (geschuetztes Leerzeichen aus Webseiten): der Kotlin-Zwilling nutzt dafuer (?U).
const COUNT_FIRST = /^([0-9]+)(?:\s*[xX])?\s+(\S.*)$/;
const COUNT_LAST = /^(.+?)\s+[xX]([0-9]+)$/;

// Passcode als Zahl ohne fuehrende Nullen ("04031928" -> "4031928"); 1..2^32-1, sonst null.
export function normalizePasscode(raw) {
  const t = String(raw).trim();
  if (!/^[0-9]{1,10}$/.test(t)) return null;
  const n = Number(t);
  return n >= 1 && n <= MAX_UINT32 ? String(n) : null;
}

// Gleiche Passcodes je Abschnitt summieren, Reihenfolge des ersten Auftretens.
function addCard(cards, passcode, section, count) {
  const hit = cards.find((c) => c.passcode === passcode && c.section === section);
  if (hit) hit.count += count;
  else cards.push({ passcode, count, section });
}

export function parseYdk(text) {
  const cards = [];
  const unresolved = [];
  let section = 'main';
  for (const raw of String(text).split(/\r?\n/)) {
    const t = raw.trim();
    if (!t) continue;
    const lower = t.toLowerCase();
    if (lower === '#main') section = 'main';
    else if (lower === '#extra') section = 'extra';
    else if (lower === '!side') section = 'side';
    else if (t.startsWith('#')) continue;
    else {
      const passcode = normalizePasscode(t);
      if (passcode) addCard(cards, passcode, section, 1);
      else unresolved.push(t);
    }
  }
  return { format: 'ydk', cards, unresolved };
}

function decodeBlock(block) {
  if (!BASE64.test(block)) return null;
  const bin = atob(block);
  if (bin.length % 4 !== 0) return null;
  const out = [];
  for (let i = 0; i < bin.length; i += 4) {
    out.push((bin.charCodeAt(i) | (bin.charCodeAt(i + 1) << 8) | (bin.charCodeAt(i + 2) << 16) | (bin.charCodeAt(i + 3) << 24)) >>> 0);
  }
  return out;
}

export function parseYdke(text) {
  const invalid = { format: 'ydke', cards: [], unresolved: [], error: YDKE_INVALID };
  const body = String(text).trim();
  if (!body.startsWith('ydke://')) return invalid;
  let parts = body.slice('ydke://'.length).replace(/\s+/g, '').split('!');
  if (parts.length === 4 && parts[3] === '') parts = parts.slice(0, 3);
  if (parts.length !== 3) return invalid;
  const cards = [];
  const unresolved = [];
  const sections = ['main', 'extra', 'side'];
  for (let s = 0; s < 3; s++) {
    const codes = decodeBlock(parts[s]);
    if (!codes) return invalid;
    for (const n of codes) {
      if (n === 0) unresolved.push('0');
      else addCard(cards, String(n), sections[s], 1);
    }
  }
  return { format: 'ydke', cards, unresolved };
}

// Kartenzeilen "3 Name", "3x Name", "3 x Name", "Name x3", "Name"; Ueberschriften Main/Extra/Side (Deck) mit ":" oder
// "(n)". Ohne Ueberschrift davor: section "unknown". Zeilen werden NICHT zusammengefasst (Rohzeile bleibt fuer die Notizen).
export function parseTextList(text) {
  const cards = [];
  const unresolved = [];
  let section = 'unknown';
  for (const raw of String(text).split(/\r?\n/)) {
    const t = raw.trim();
    if (!t || t.startsWith('#') || t.startsWith('//')) continue;
    const heading = HEADING.exec(t);
    if (heading) { section = heading[1].toLowerCase(); continue; }
    let count = 1;
    let name = t;
    const first = COUNT_FIRST.exec(t);
    const last = first ? null : COUNT_LAST.exec(t);
    if (first) { count = Number(first[1]); name = first[2]; }
    else if (last) { name = last[1]; count = Number(last[2]); }
    if (!(count >= 1 && count <= 99)) { unresolved.push(t); continue; }
    cards.push({ name: name.trim(), count, section, line: t });
  }
  return { format: 'text', cards, unresolved };
}

export function detectFormat(text) {
  const body = String(text).trim();
  if (body.startsWith('ydke://')) return 'ydke';
  const lines = body.split(/\r?\n/).map((l) => l.trim().toLowerCase());
  if (lines.includes('#main') || lines.includes('!side')) return 'ydk';
  return 'text';
}

// Einstieg fuer Einfuegen/Teilen/YDK-Datei: erkennt (oder nimmt das vorgegebene Format), liest, und meldet
// "Keine Deckliste erkannt", wenn keine einzige Karte gelesen wurde.
export function parseDeckText(text, format = detectFormat(text)) {
  const parsed = format === 'ydke' ? parseYdke(text) : format === 'ydk' ? parseYdk(text) : parseTextList(text);
  if (parsed.error) return parsed;
  if (parsed.cards.length === 0) return { ...parsed, error: NOTHING_RECOGNIZED };
  return parsed;
}

const live = (entries) => (entries || []).filter((e) => e && e.count > 0);
const ofSection = (entries, section) => live(entries).filter((e) => e.section === section);

// entries: [{ passcode, name?, count, section }] (mehrere Zeilen je Passcode erlaubt).
export function buildYdk(entries) {
  const lines = [YDK_HEADER];
  for (const [head, section] of [['#main', 'main'], ['#extra', 'extra'], ['!side', 'side']]) {
    lines.push(head);
    for (const e of ofSection(entries, section)) for (let i = 0; i < e.count; i++) lines.push(String(e.passcode));
  }
  return `${lines.join('\n')}\n`;
}

function encodeBlock(entries) {
  let bin = '';
  for (const e of entries) {
    const n = Number(e.passcode) >>> 0;
    const four = String.fromCharCode(n & 255, (n >>> 8) & 255, (n >>> 16) & 255, (n >>> 24) & 255);
    for (let i = 0; i < e.count; i++) bin += four;
  }
  return btoa(bin);
}

export function buildYdke(entries) {
  return `ydke://${['main', 'extra', 'side'].map((s) => `${encodeBlock(ofSection(entries, s))}!`).join('')}`;
}

// "Main Deck" / "3 Name" …; je Abschnitt nach Passcode summiert (Reihenfolge des ersten Auftretens), leere Abschnitte
// entfallen, Name faellt auf den Passcode zurueck. Zeilen mit "\n" verbunden, ohne Zeilenumbruch am Ende.
export function buildTextList(entries) {
  const lines = [];
  for (const [head, section] of [['Main Deck', 'main'], ['Extra Deck', 'extra'], ['Side Deck', 'side']]) {
    const summed = [];
    for (const e of ofSection(entries, section)) {
      const hit = summed.find((x) => x.passcode === String(e.passcode));
      if (hit) hit.count += e.count;
      else summed.push({ passcode: String(e.passcode), name: e.name || String(e.passcode), count: e.count });
    }
    if (summed.length === 0) continue;
    lines.push(head);
    for (const x of summed) lines.push(`${x.count} ${x.name}`);
  }
  return lines.join('\n');
}

File: android/app/src/main/java/com/example/yugiohscanner/ml/DeckFormats.kt
package com.example.yugiohscanner.ml

/** Gelesene Karte: YDK/YDKE mit [passcode], Textliste mit [name] und Rohzeile [line]. section: main|extra|side|unknown. */
data class ParsedCard(
    val passcode: String?,
    val name: String?,
    val count: Int,
    val section: String,
    val line: String?,
)

data class ParsedDeck(val format: String, val cards: List<ParsedCard>, val unresolved: List<String>, val error: String? = null)

/** Eine Deckzeile zum Schreiben (mehrere Zeilen je Passcode erlaubt). */
data class DeckEntry(val passcode: String, val name: String?, val count: Int, val section: String)

/**
 * Spec E2 §3 -- Decklisten-Formate YDK, YDKE und Textliste: lesen, schreiben, erkennen.
 * ZWILLING: desktop/src/utils/deckFormats.js. Beide laufen gegen docs/fixtures/decks/formats.json.
 * Wer eine Seite aendert, aendert beide.
 */
object DeckFormats {
    const val YDKE_INVALID = "Kein gültiger YDKE-Link"
    const val NOTHING_RECOGNIZED = "Keine Deckliste erkannt"
    const val YDK_HEADER = "#created by YGO Card Manager"

    private const val MAX_UINT32 = 4294967295L
    // Standard-Base64 mit Auffuellung; dasselbe Muster wie der JS-Zwilling (java.util.Base64 ist nachsichtiger).
    private val BASE64 = Regex("^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$")
    // (?U): \s wie im JS-Zwilling auch fuer geschuetzte Leerzeichen aus Webseiten; Ziffern bewusst [0-9].
    private val HEADING = Regex("(?U)^(main|extra|side)(?:\\s+deck)?\\s*(?::|\\([0-9]+\\))?$", RegexOption.IGNORE_CASE)
    private val COUNT_FIRST = Regex("(?U)^([0-9]+)(?:\\s*[xX])?\\s+(\\S.*)$")
    private val COUNT_LAST = Regex("(?U)^(.+?)\\s+[xX]([0-9]+)$")
    private val LINE_BREAK = Regex("\\r?\\n")
    private val WHITESPACE = Regex("(?U)\\s+")
    private val PASSCODE = Regex("^[0-9]{1,10}$")
    private val SECTIONS = listOf("main", "extra", "side")

    /** Passcode als Zahl ohne fuehrende Nullen; 1..2^32-1, sonst null. */
    fun normalizePasscode(raw: String): String? {
        val t = raw.trim()
        if (!PASSCODE.matches(t)) return null
        val n = t.toLong()
        return if (n in 1..MAX_UINT32) n.toString() else null
    }

    private fun MutableList<ParsedCard>.addPasscode(passcode: String, section: String) {
        val i = indexOfFirst { it.passcode == passcode && it.section == section }
        if (i >= 0) this[i] = this[i].copy(count = this[i].count + 1)
        else add(ParsedCard(passcode, null, 1, section, null))
    }

    fun parseYdk(text: String): ParsedDeck {
        val cards = mutableListOf<ParsedCard>()
        val unresolved = mutableListOf<String>()
        var section = "main"
        for (raw in text.split(LINE_BREAK)) {
            val t = raw.trim()
            if (t.isEmpty()) continue
            when (t.lowercase()) {
                "#main" -> section = "main"
                "#extra" -> section = "extra"
                "!side" -> section = "side"
                else -> if (!t.startsWith("#")) {
                    val passcode = normalizePasscode(t)
                    if (passcode != null) cards.addPasscode(passcode, section) else unresolved.add(t)
                }
            }
        }
        return ParsedDeck("ydk", cards, unresolved)
    }

    private fun decodeBlock(block: String): List<Long>? {
        if (!BASE64.matches(block)) return null
        val bytes = java.util.Base64.getDecoder().decode(block)
        if (bytes.size % 4 != 0) return null
        return (bytes.indices step 4).map { i ->
            (bytes[i].toLong() and 0xFF) or ((bytes[i + 1].toLong() and 0xFF) shl 8) or
                ((bytes[i + 2].toLong() and 0xFF) shl 16) or ((bytes[i + 3].toLong() and 0xFF) shl 24)
        }
    }

    fun parseYdke(text: String): ParsedDeck {
        val invalid = ParsedDeck("ydke", emptyList(), emptyList(), YDKE_INVALID)
        val body = text.trim()
        if (!body.startsWith("ydke://")) return invalid
        var parts = body.removePrefix("ydke://").replace(WHITESPACE, "").split("!")
        if (parts.size == 4 && parts[3].isEmpty()) parts = parts.take(3)
        if (parts.size != 3) return invalid
        val cards = mutableListOf<ParsedCard>()
        val unresolved = mutableListOf<String>()
        for (s in 0 until 3) {
            val codes = decodeBlock(parts[s]) ?: return invalid
            for (n in codes) {
                if (n == 0L) unresolved.add("0") else cards.addPasscode(n.toString(), SECTIONS[s])
            }
        }
        return ParsedDeck("ydke", cards, unresolved)
    }

    fun parseTextList(text: String): ParsedDeck {
        val cards = mutableListOf<ParsedCard>()
        val unresolved = mutableListOf<String>()
        var section = "unknown"
        for (raw in text.split(LINE_BREAK)) {
            val t = raw.trim()
            if (t.isEmpty() || t.startsWith("#") || t.startsWith("//")) continue
            val heading = HEADING.find(t)
            if (heading != null) { section = heading.groupValues[1].lowercase(); continue }
            var count = 1
            var name = t
            val first = COUNT_FIRST.find(t)
            val last = if (first == null) COUNT_LAST.find(t) else null
            if (first != null) { count = first.groupValues[1].toIntOrNull() ?: 0; name = first.groupValues[2] }
            else if (last != null) { name = last.groupValues[1]; count = last.groupValues[2].toIntOrNull() ?: 0 }
            if (count !in 1..99) { unresolved.add(t); continue }
            cards.add(ParsedCard(null, name.trim(), count, section, t))
        }
        return ParsedDeck("text", cards, unresolved)
    }

    fun detectFormat(text: String): String {
        val body = text.trim()
        if (body.startsWith("ydke://")) return "ydke"
        val lines = body.split(LINE_BREAK).map { it.trim().lowercase() }
        return if ("#main" in lines || "!side" in lines) "ydk" else "text"
    }

    /** Einstieg fuer Einfuegen/Teilen: erkennt (oder nimmt [format]), liest, "Keine Deckliste erkannt" ohne Karte. */
    fun parseDeckText(text: String, format: String = detectFormat(text)): ParsedDeck {
        val parsed = when (format) {
            "ydke" -> parseYdke(text)
            "ydk" -> parseYdk(text)
            else -> parseTextList(text)
        }
        if (parsed.error != null) return parsed
        return if (parsed.cards.isEmpty()) parsed.copy(error = NOTHING_RECOGNIZED) else parsed
    }

    private fun ofSection(entries: List<DeckEntry>, section: String) = entries.filter { it.count > 0 && it.section == section }

    fun buildYdk(entries: List<DeckEntry>): String {
        val lines = mutableListOf(YDK_HEADER)
        for ((head, section) in listOf("#main" to "main", "#extra" to "extra", "!side" to "side")) {
            lines.add(head)
            for (e in ofSection(entries, section)) repeat(e.count) { lines.add(e.passcode) }
        }
        return lines.joinToString("\n") + "\n"
    }

    private fun encodeBlock(entries: List<DeckEntry>): String {
        val out = java.io.ByteArrayOutputStream()
        for (e in entries) {
            val n = e.passcode.toLong()
            repeat(e.count) {
                out.write((n and 0xFF).toInt()); out.write(((n shr 8) and 0xFF).toInt())
                out.write(((n shr 16) and 0xFF).toInt()); out.write(((n shr 24) and 0xFF).toInt())
            }
        }
        return java.util.Base64.getEncoder().encodeToString(out.toByteArray())
    }

    fun buildYdke(entries: List<DeckEntry>): String =
        "ydke://" + SECTIONS.joinToString("") { encodeBlock(ofSection(entries, it)) + "!" }

    fun buildTextList(entries: List<DeckEntry>): String {
        val lines = mutableListOf<String>()
        for ((head, section) in listOf("Main Deck" to "main", "Extra Deck" to "extra", "Side Deck" to "side")) {
            val summed = LinkedHashMap<String, Pair<String, Int>>()
            for (e in ofSection(entries, section)) {
                val prev = summed[e.passcode]
                summed[e.passcode] = if (prev == null) (e.name?.takeIf { it.isNotEmpty() } ?: e.passcode) to e.count
                else prev.first to prev.second + e.count
            }
            if (summed.isEmpty()) continue
            lines.add(head)
            for ((name, count) in summed.values) lines.add("$count $name")
        }
        return lines.joinToString("\n")
    }
}
